### ISSUE TITLE
Provide more configuration-cached compliant listeners

### ISSUE THREAD

**Issue by celcius112** [gradle/gradle#14860]
Hello everyone

We are testing the migration of our plugins to be configuration-cache compliant. One blocking point we have is the migration of our `BuildListener` and `TaskExecutionListener`. As described in [the doc](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:requirements:build_listeners) , we can use a `OperationCompletionListener` with a 
`BuildEventsListenerRegistry` to listen to task completion event. This works fine. However, our current listeners also expect to receive an event when a task is started as well as when a build finishes. More globally, it seems like only `TaskExecutionListener#afterExecute` has a clear migration path according to the doc.

To sum it up, it would be great to have access to more listeners in a configuration-cached compliant Gradle environment.

**Comment by adammurdoch** [comment]
Can you give some more details about what you used these listeners for? For example, what does the listener do with the start event? Or which events from `BuildListener` do you use and what does the listener do with these events?

**Comment by celcius112** [comment]
Hello @adammurdoch,

Right now we do nothing fancy. We log a timestamp when a task starts (`TaskExecutionListener#beforeExecute`), and generate a report at the end of the build containing the execution time of the longest tasks (using `TaskExecutionListener#afterExecute`  and `BuildListener#buildFinished`). 

In the end, we only require the task's path when consuming the events.

**Comment by celcius112** [comment]
Good idea for the AutoCloseable, I'll try that. We are however still in need of a way to log a timestamp as a task starts, a kind of `OperationStartedListener`. I see that there exists the counterpart of `TaskFinishEvent`, `TaskStartEvent`, but no corresponding listener and `BuildEventsListenerRegistry` subscriber.

**Comment by eskatos** [comment]
You can get the start timestamp from the finish event. Or do you really need it before the task starts?

**Comment by celcius112** [comment]
Actually, as soon as the task starts. This has become useful in the CI when we have arrived on a build with a task being stuck for some reason. It helps us to quickly comprehend that there is some issue going on. Waiting for the starting timestamp in a `OperationCompletionListener` requires to wait for the task to be finished.

However, reporting the time taken for each task is still more important to us. The `AutoCloseable` proposition is great and we can live only with that if necessary.

**Comment by Vampire** [comment]
> A build service that implements OperationCompletionListener and is registered with the BuildEventsListenerRegistry will be started at the beginning of the build and if it implements AutoCloseable its close() function will be called at the end of the build.

"at the beginning of the build" seems to be a bit misleading.
It actually means after the first task was executed.
I also second having more  listeners possible.
I for example up to now print in the configuration phase some TeamCity service messages that set various values, for example the version with generated build label.

This should be done as soon as possible.
Doing it during configuration was fine but will not work of course with the configuration cache.
But doing it in the constructor of a build service will only do it after the first task finished, which is actually a bit late.
It would be nice if there were a way to do it right after the confiugration phase.
Or maybe even selectively before or after the configuration phase.

**Comment by Vampire** [comment]
And as a side-note for people having similar use-cases, if you all `get()` on the `Provider<YourBuildService>` during configuration time, it will immediately be created, so the constructor code is executed as soon as that code is reached. So if the configuration is executed, the code is immediately executed, if the configuration comes from the cache, it is executed as soon as the first task is finished. And if this is also on a CI server and there is no configuration cache available, it will always be done during configuration phase.

**Comment by breskeby** [comment]
In the elasticsearch build we have the notion of a `ClusterAware` task that uses an elasticsearch cluster within that task. Those tasks implement a common `TestClustersAware`  interface. Those tasks requires some additional wiring cleanup that we do before and after those tasks. Currently we use a TestActionListener and a TaskExecutionListener to set those up. I don't see a clean good way to migrate this away from using those listeners and would love to see a configuration cache compliant solution.

**Comment by eskatos** [comment]
@breskeby, it sounds like a `BuildService` that those tasks would use to setup/cleanup things should do the trick. The tasks would have to call the build service around their task action though. Please reach out in the `#configuration-cache` channel in the community slack if you want to discuss details.

**Comment by ThanosFisherman** [comment]
Is there a way to determine whether a build failed or succeeded? I'm trying to migrate my `BuildListener` logic into a cached-compliant one. The trick with `AutoCloseable` helped me execute my code at the very end of the build but I don't know how to check whether the build succeeded or failed like I do with the `buildFinished` overriden method of `BuildListener`

```kotlin
    override fun buildFinished(result: BuildResult) {

        if (result.failure != null) {
            // build failure
        } else {
            // build success!
        }
    }
```

**Comment by anthonymonori** [comment]
@ThanosFisherman on a BuildService you can implement the OperationCompletionListener and look at each individual task result and store if there was any failure:

```
    override fun onFinish(event: FinishEvent?) {
        event?.let {
            if (it.result is TaskFailureResult) {
                // Store failure and/or reason
            }
        }
    }
 ```
 
There's no such easy way to just get the output since given how BuildService is executed differently to how BuildListener worked.

**Comment by eskatos** [comment]
Related 
- https://github.com/gradle/gradle/issues/17659
